src/librustc_typeck/check/method/confirm.rs

```diff
@@ -332,7 +332,9 @@ impl<'a, 'gcx, 'tcx> ConfirmContext<'a, 'gcx, 'tcx> {
                 parent_substs.type_at(i)
             } else if let Some(ast_ty)
                 = provided.as_ref().and_then(|p| {
-                    p.types.get(i - parent_substs.len() - method_generics.regions.len())
+                    let idx =
+                        i - parent_substs.len() - method_generics.param_counts().lifetimes;
```

Comment:
Note that this is accidentally quadratic, `param_counts` shouldn't be called within this closure.